compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionUtils.cpp

```diff
@@ -56,6 +63,59 @@ static bool canInsOperandTieWithOutsOperand(OpOperand *insOperand) {
   return llvm::any_of(linalgOp.getOutputOperands(), canTieWithOutsOperand);
 }
 
+/// Checks if a linalg op is a simple reduction of the innermost dimensions
+/// with identity map for the input.
+static bool isSimpleReduction(linalg::LinalgOp linalgOp) {
+  SmallVector<Operation *, 4> combinerOps;
+
+  if (linalgOp.getNumOutputs() != 1) return false;
+
+  // Check if the result dims are d0, d1, ..., which means the reduction is done
+  // in the innermost dimensions.
+  auto output = linalgOp.getOutputOperand(0);
+  auto outputIndexingMap = linalgOp.getTiedIndexingMap(output);
+  for (const auto &en : llvm::enumerate(outputIndexingMap.getResults())) {
+    auto expr = en.value();
+    if (auto dim = expr.dyn_cast<AffineDimExpr>()) {
+      if (dim.getPosition() != en.index()) return false;
+    } else {
+      return false;
+    }
+  }
+
+  if (!matchReduction(linalgOp.getRegionOutputArgs(), 0, combinerOps) ||
+      combinerOps.size() != 1) {
+    return false;
+  }
+
+  return true;
+}
+
+/// Check if the use of operand is a reduction -> broadcast -> elementwise.
+static bool isReductionBroadcastElementwise(OpOperand *operand) {
+  auto producer = operand->get().getDefiningOp<linalg::LinalgOp>();
+  auto consumer = dyn_cast<linalg::LinalgOp>(operand->getOwner());
+  if (!producer || !consumer) return false;
+
+  // Check if the producer is a simple reduction.
+  if (!isSimpleReduction(producer)) return false;
+
+  // Check if the reduction is broadcasted back for the elementwise op.
+  auto output = producer.getOutputOperand(0);
+  auto outputIndexingMap = producer.getTiedIndexingMap(output);
+  auto inputIndexingMap = consumer.getTiedIndexingMap(operand);
+  if (outputIndexingMap != inputIndexingMap) return false;
+
+  // Check if the consumer is an elementwise op.
+  if (!isElementwise(consumer)) return false;
+
+  // Check if there is no output transpose.
+  if (producer->getOperand(0).getType() != consumer->getResult(0).getType())
```

Comment:
I don’t think this actually check that there is no transpose. You could have the same transpose version of type be the same type (ex: <4x4xf32> -> <4x4xf32>). I don’t think you can know just based on the type.

Reply:
Good point. I will revise the code. Thanks

Reply:
Done. Thanks